bliss/surveys/decals.py

```diff
@@ -1,13 +1,161 @@
 from pathlib import Path
 from typing import Tuple
 
+import numpy as np
+import pytorch_lightning as pl
 import torch
+from astropy.io import fits
 from astropy.table import Table
 from astropy.wcs import WCS
+from numpy.core.defchararray import startswith
+from pyvo.dal import sia
+from torch.utils.data import DataLoader, Dataset
 
 from bliss.catalog import FullCatalog, SourceType
-from bliss.surveys.sdss import column_to_tensor
-from bliss.utils.download_utils import download_git_lfs_file
+from bliss.surveys.sdss import column_to_tensor, prepare_batch, prepare_image
+from bliss.utils.download_utils import download_file_to_dst
+
+
+class DarkEnergyCameraLegacySurvey(pl.LightningDataModule, Dataset):
+    PIX_SCALE = 0.262  # arcsec/pixel
+    PIX_SIZE = 3600  # arcsec/degree
+
+    def __init__(
+        self,
+        decals_dir="data/decals",
+        ra=335,
+        dec=0,
+        width=2400,
+        height=1489,
+        bands=("g", "r", "i", "z"),
+        predict_device=None,
+        predict_crop=None,
+    ):
+        super().__init__()
+
+        self.decals_path = Path(decals_dir)
+        self.ra = ra
+        self.dec = dec
+        self.width = width
+        self.height = height
+        self.bands = bands
+
+        self.downloader = DecalsDownloader(ra, dec, width, height, self.decals_path)
+
+        self.items = []
+        self.prepare_data()
+        assert self.items is not None, "No data found even after prepare_data()."
+
+        self.predict_device = predict_device
+        self.predict_crop = predict_crop
+
+    def prepare_data(self):
+        for band in self.bands:
+            self.brick_name = self.downloader.download_cutout(band)
+        assert self.brick_name is not None, "No brick ID found even after prepare_data()."
+        self.downloader.download_catalog(self.brick_name)
+
+    def __len__(self):
+        return 1
+
+    def __getitem__(self, idx):
+        return self.get_from_disk(idx)
+
+    def get_from_disk(self, idx):
+        cutout_list = []
+        for band in self.bands:
+            cutout_list.append(self.read_cutout_for_band(band))
+
+        ret = {}
+        for k in cutout_list[0]:
+            data_per_band = [cutout[k] for cutout in cutout_list]
+            if isinstance(data_per_band[0], np.ndarray):
+                ret[k] = np.stack(data_per_band)
+            else:
+                ret[k] = data_per_band
+        return ret
+
+    def read_cutout_for_band(self, band):
+        cutout = fits.open(self.decals_path / f"cutout_{self.ra}_{self.dec}_{band}.fits")
+        image = cutout[0].data  # pylint: disable=no-member
+        hr = cutout[0].header  # pylint: disable=no-member
+        wcs = WCS(hr)
+        return {
+            "image": image,
+            "background": np.zeros_like(image),  # TODO: find a way to get background
+            "wcs": wcs,
```

Comment:
Do we read in the PSF somewhere too? I think we'd need the DECaLS PSFs to generate realistic training data.

Reply:
DECaLS data generation - punting to after #853.

Reply:
tracked by #871